File: cs50/problem_set_1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)

{

    long long card_number;
    int length = 0; // keeps count of the lenght of the card
    int sum = 0;    // keeps track of checksum

    do
    {

        card_number = get_long_long("Card Number: ");
    }

    while (card_number < 0);

    // every other number. starting from second-to-last

    for (long long i = card_number / 10; i > 0; i /= 100)
    {

        int digit = (i % 10) * 2;
        sum += digit / 10 + digit % 10;
        length++;
    }
    //

    // rest of the numbers not multiplied by 2
    for (long long i = card_number; i > 0; i /= 100)
    {
        int digit = i % 10;
        sum += digit;
        length++;
    }
    //

    // checksum
    if (sum % 10 == 0)
    {

        if (length == 15 && (card_number / 10000000000000 == 34 || card_number / 10000000000000 == 37))
        {
            printf("AMEX\n");
        }
        else if (length == 16 && (card_number / 100000000000000 > 50 && card_number / 100000000000000 < 56))
        {
            printf("MASTERCARD\n");
        }
        else if ((length == 13 || length == 16) && (card_number / 1000000000000 == 4 || card_number / 1000000000000000 == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
/// checksum

File: cs50/problem_set_6/sentimental-credit/credit.py

def main():
    credit = get_credit()
    checksum = lunh(credit)
    is_valid(checksum, credit)


def get_credit():
    return (input("Credit card number: ").strip())



def lunh(card):
    total_sum = 0
    for i in range(len(card)-2, -1, -2):
        digit = int(card[i]) * 2
        if digit > 9:
            digit = (digit % 10) + (digit // 10)
        total_sum += digit

    for i in range(len(card) - 1, -1, -2):
        total_sum += int(card[i]) % 10
    return int(total_sum)


def is_valid(card, credit):
    if card % 10 == 0:
        if len(credit) == 15 and (credit[:2] == "34" or credit[:2] == '37'):
            print("AMEX")
        elif len(credit) == 16 and (int(credit[:2]) > 50 and int(credit[:2]) < 56):
            print("MASTERCARD")
        elif (len(credit) == 13 or len(credit) == 16) and credit[0] == "4":
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


main()
